tests/utils.py

```diff
@@ -197,3 +197,40 @@ def create_proxy_block_file(blocked_domains: List[str]):
         file.flush()
 
         yield file.name
+
+
+@contextlib.contextmanager
+def slow_app(host: str, port: int):
+    stop = False
+
+    def run():
+        nonlocal stop
+        sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
+        sock.settimeout(0.1)
+
+        try:
+            sock.bind((host, port))
+            sock.listen(10)
+            print(f"Successfully bind slow_app on {host}:{port}")
+
+            while not stop:
+                conn = None
+                try:
+                    conn, _ = sock.accept()
+                    time.sleep(0.5)
+                except socket.timeout:
+                    pass
+                finally:
+                    if conn is not None:
+                        conn.close()
+        finally:
+            sock.close()
+
+    thread = threading.Thread(target=run)
+
+    try:
+        thread.start()
+        yield
+        stop = True
+    finally:
+        thread.join()
```

Comment:
Should this be…?

```suggestion
    finally:
        stop = True
        thread.join()
```

(If an exception occurs around the `yield`, we _do_ want `stop` to be set, otherwise we'd have a deadlock as the thread would never stop.)